gradle-baseline-java/src/main/java/com/palantir/baseline/plugins/javaversions/BaselineJavaVersions.java

```diff
@@ -75,43 +69,38 @@ public void apply(Project project) {
+            log.info(
+                    "{} is considered a library because it has been overridden with library = true",
                     project.getDisplayName());
             return libraryOverride.get();
         }
 
         for (String plugin : LIBRARY_PLUGINS) {
             if (project.getPluginManager().hasPlugin(plugin)) {
-                log.debug(
-                        "Project '{}' is considered a library because the '{}' plugin is applied",
+                log.info(
+                        "{} is considered a library because the '{}' plugin is applied",
                         project.getDisplayName(),
                         plugin);
                 return true;
             }
         }
 
-        for (String plugin : DISTRIBUTION_PLUGINS) {
-            if (project.getPluginManager().hasPlugin(plugin)) {
-                log.debug(
-                        "Project '{}' is considered a distribution because the '{}' plugin is applied",
-                        project.getDisplayName(),
-                        plugin);
-                return false;
-            }
-        }
-
-        PublishingExtension publishing = project.getExtensions().findByType(PublishingExtension.class);
-        if (publishing == null) {
-            log.debug(
-                    "Project '{}' is considered a distribution, not a library, because "
-                            + "it doesn't define any publishing extensions",
-                    project.getDisplayName());
-            return false;
-        }
-
-        // Better to be conservative with the java version rather than release something that is too high to be used.
-        log.debug("Project '{}' is considered a library as no other conditions matched", project.getDisplayName());
-        return true;
+        // The whole reason for having libraryTarget vs distributionTarget is so that we can allow distributions
+        // (ie Java code that isn't published and only used in that service) to use newer source features and
+        // class files, allowing everyone to support the newer compilation level *before* libraries they depend on
+        // start requiring the newer compilation level. We want to upgrade distributionTarget first, wait for every
+        // repo to support the new compilation level, then allow libraries to be published with the new level.
+        //
+        // So the assumptions here are that (at least for Palantir code):
+        //   1. Only individual published jars matter for spreading higher compilation level requirements between repos
+        //      when we upgrade Java versions:
+        //        a) Jars inside dists/containers are not relevant as they will normally have JDK packaged with them.
+        //        b) Shaded/fat jars will still need to be published to have any kind of effect.
+        //   2. The *only* way to publish an individual jar is to use one of the above library plugins.
```

Comment:
Do we care about non-Palantir publishing plugins? I'm guessing not (especially since there's not a Gradle "JAR publishing" plugin, there's just `maven-publish`), but just wanted to confirm we're okay with this tradeoff.

Reply:
I guess we might want the gradle plugin publish plugin too. I can't think of anything else (that we use at Palantir) that should be publishing individual jars.

Eventually (once this is rewritten) I want this is move back to checking if there's any publication that has a jar artifact, but we can't do this without the rewrite listed in the description (due to [this](https://github.com/palantir/gradle-baseline/pull/2820))